core/trino-main/src/main/java/io/trino/metadata/DiscoveryNodeManager.java

```diff
@@ -208,81 +214,86 @@ public void refreshNodes()
-        }
-
-        if (allNodes != null) {
-            // log node that are no longer active (but not shutting down)
-            SetView<InternalNode> missingNodes = difference(allNodes.getActiveNodes(), Sets.union(activeNodesBuilder.build(), shuttingDownNodesBuilder.build()));
-            for (InternalNode missingNode : missingNodes) {
-                log.info("Previously active node is missing: %s (last seen at %s)", missingNode.getNodeIdentifier(), missingNode.getHost());
+            if (allNodes != null) {
+                // log node that are no longer active (but not shutting down)
+                SetView<InternalNode> missingNodes = difference(allNodes.getActiveNodes(), Sets.union(activeNodesBuilder.build(), shuttingDownNodesBuilder.build()));
+                for (InternalNode missingNode : missingNodes) {
+                    log.info("Previously active node is missing: %s (last seen at %s)", missingNode.getNodeIdentifier(), missingNode.getHost());
+                }
             }
-        }
 
-        // nodes by catalog handle changes anytime a node adds or removes a catalog (note: this is not part of the listener system)
-        if (!allCatalogsOnAllNodes) {
-            activeNodesByCatalogHandle = Optional.of(byCatalogHandleBuilder.build());
-        }
+            // nodes by catalog handle changes anytime a node adds or removes a catalog (note: this is not part of the listener system)
+            if (!allCatalogsOnAllNodes) {
+                activeNodesByCatalogHandle = Optional.of(byCatalogHandleBuilder.build());
+            }
 
-        AllNodes allNodes = new AllNodes(activeNodesBuilder.build(), inactiveNodesBuilder.build(), shuttingDownNodesBuilder.build(), coordinatorsBuilder.build());
-        // only update if all nodes actually changed (note: this does not include the connectors registered with the nodes)
-        if (!allNodes.equals(this.allNodes)) {
-            // assign allNodes to a local variable for use in the callback below
-            this.allNodes = allNodes;
-            coordinators = coordinatorsBuilder.build();
+            AllNodes allNodes = new AllNodes(activeNodesBuilder.build(), inactiveNodesBuilder.build(), shuttingDownNodesBuilder.build(), coordinatorsBuilder.build());
+            // only update if all nodes actually changed (note: this does not include the connectors registered with the nodes)
+            if (!allNodes.equals(this.allNodes)) {
+                // assign allNodes to a local variable for use in the callback below
+                this.allNodes = allNodes;
+                coordinators = coordinatorsBuilder.build();
 
-            // notify listeners
-            List<Consumer<AllNodes>> listeners = ImmutableList.copyOf(this.listeners);
-            nodeStateEventExecutor.submit(() -> listeners.forEach(listener -> listener.accept(allNodes)));
+                // notify listeners
+                List<Consumer<AllNodes>> listeners = ImmutableList.copyOf(this.listeners);
```

Comment:
This also needs to synchronize on accessing `listeners`. Alternatively you could:
- switch all accesses of `listeners` to using the read write lock and updated the `@GuardedBy` annotation.
- still `synchronize (this)` on accessing listeners, but do so after releasing the `allNodesLock` to avoid potential deadlock scenarios